src/rgw/rgw_d3n_datacache.h

```diff
@@ -0,0 +1,366 @@
+    o->lru_next = head;
+    o->lru_prev = nullptr;
+    if (head) {
+      head->lru_prev = o;
+    } else {
+      tail = o;
+    }
+    head = o;
+  }
+  void lru_insert_tail(struct D3nChunkDataInfo* o) {
+    o->lru_next = nullptr;
+    o->lru_prev = tail;
+    if (tail) {
+      tail->lru_next = o;
+    } else {
+      head = o;
+    }
+    tail = o;
+  }
+
+  void lru_remove(struct D3nChunkDataInfo* o) {
+    if (o->lru_next)
+      o->lru_next->lru_prev = o->lru_prev;
+    else
+      tail = o->lru_prev;
+    if (o->lru_prev)
+      o->lru_prev->lru_next = o->lru_next;
+    else
+      head = o->lru_next;
+    o->lru_next = o->lru_prev = nullptr;
+  }
+};
+
+
+template <class T>
+class D3nRGWDataCache : public T
+{
+
+  D3nDataCache d3n_data_cache;
+
+public:
+  D3nRGWDataCache() {}
+
+  int init_rados() override {
+    int ret;
+    d3n_data_cache.init(T::cct);
+    ret = T::init_rados();
+    if (ret < 0)
+      return ret;
+
+    return 0;
+  }
+
+  int flush_read_list(struct get_obj_data* d);
+  int get_obj_iterate_cb(const rgw_raw_obj& read_obj, off_t obj_ofs,
+                         off_t read_ofs, off_t len, bool is_head_obj,
+                         RGWObjState *astate, void *arg) override;
+};
+
+
+template<typename T>
+int D3nRGWDataCache<T>::flush_read_list(struct get_obj_data* d) {
+
+  d->d3n_datacache_lock.lock();
+  std::list<bufferlist> l;
+  l.swap(d->d3n_read_list);
+  d->d3n_read_list.clear();
+  d->d3n_datacache_lock.unlock();
+
+  int r = 0;
+
+  std::string oid;
+  std::list<bufferlist>::iterator iter;
+  for (iter = l.begin(); iter != l.end(); ++iter) {
+    bufferlist& bl = *iter;
+    oid = d->d3n_get_pending_oid();
+    if(oid.empty()) {
+      lsubdout(g_ceph_context, rgw, 0) << "ERROR: flush_read_list(): d3n_get_pending_oid() returned empty oid" << dendl;
+      r = -ENOENT;
+      break;
+    }
+    if (bl.length() <= 0x400000)
+      d3n_data_cache.put(bl, bl.length(), oid);
+  }
+
+  return r;
+}
+
+template<typename T>
+int D3nRGWDataCache<T>::get_obj_iterate_cb(const rgw_raw_obj& read_obj, off_t obj_ofs,
+                                 off_t read_ofs, off_t len, bool is_head_obj,
+                                 RGWObjState *astate, void *arg) {
+
+  librados::ObjectReadOperation op;
+  struct get_obj_data* d = static_cast<struct get_obj_data*>(arg);
+  string oid, key;
+
+  int r = 0;
+
+  if (is_head_obj) {
+    // only when reading from the head object do we need to do the atomic test
+    r = T::append_atomic_test(astate, op);
+    if (r < 0)
+      return r;
+
+    if (astate && obj_ofs < astate->data.length()) {
+      unsigned chunk_len = std::min((uint64_t)astate->data.length() - obj_ofs, (uint64_t)len);
+
+      r = d->client_cb->handle_data(astate->data, obj_ofs, chunk_len);
+      if (r < 0)
+        return r;
+
+      len -= chunk_len;
+      d->offset += chunk_len;
+      read_ofs += chunk_len;
+      obj_ofs += chunk_len;
+      if (!len)
+        return 0;
+    }
+  }
```

Comment:
i'd really like to avoid duplicating all this stuff between `RGWRados` and `D3nRGWDataCache<T>`, especially because https://github.com/ceph/ceph/pull/36144 is making changes to prefetching here

maybe we can break `get_obj_iterate_cb()` into two pieces; keep everything that's common there, but add a new virtual function that submits the read?
```c++
virtual int submit_chunk_read(ObjectOperation&&, get_obj_data* d, obj_ofs, read_ofs, len);
```

Reply:
may I ask please to postpone this refactoring after the merge of the initial D3N